src/Data/Zebra/Word/Internal.hs

```diff
@@ -0,0 +1,2906 @@
+{-# LANGUAGE BangPatterns
+           , PatternSynonyms
+           , ViewPatterns
+           , UnboxedTuples
+           , UnboxedSums #-}
+
+module Data.Zebra.Word.Internal
+  ( Color (..)
+  , Zebra (Mono, ..)
+
+  , Data.Zebra.Word.Internal.lookup
+  , lookupL
+  , findL
+  , lookupR
+  , findR
+
+  , Range (..)
+
+  , monoL
+  , monoR
+  , monoRange
+
+  , unsafeMonoRange
+
+  , size
+
+  , sizeL
+  , sizeR
+  , sizeRange
+
+  , unsafeSize
+  , unsafeSizeL
+  , unsafeSizeR
+  , unsafeSizeRange
+
+  , fillL
+  , fillR
+  , fillRange
+
+  , unsafeFillL
+  , unsafeFillRange
+
+  , Data.Zebra.Word.Internal.foldl
+  , foldlL
+  , foldlR
+  , foldlRange
+  , unsafeFoldlRange
+
+  , Data.Zebra.Word.Internal.foldr
+  , foldrL
+  , foldrR
+  , foldrRange
+  , unsafeFoldrRange
+
+  , Data.Zebra.Word.Internal.foldl'
+  , foldlL'
+  , foldlR'
+  , foldlRange'
+  , unsafeFoldlRange'
+
+  , Data.Zebra.Word.Internal.foldr'
+  , foldrL'
+  , foldrR'
+  , foldrRange'
+  , unsafeFoldrRange'
+
+  , Data.Zebra.Word.Internal.complement
+
+  , union
+  , disjoint
+  , intersection
+
+  , difference
+  , symmetricDifference
+
+  , Data.Zebra.Word.Internal.compare
+  ) where
+
+import           Radix.Common (PartialOrdering (..), order)
+import           Radix.Word.Common
+import           Radix.Word.Foundation
+
+import           Data.Bits
+import           Numeric.Natural
+
+
+
+-- | Space partition colors.
+data Color = Black
+           | White
+             deriving (Show, Eq)
+
+invert :: Color -> (# Color #)
```

Comment:
It seems that returning singleton strict tuples in cases like this could be avoided at no cost? Do these singleton tuples serve some purpose, in particular for the `invert` function?

Reply:
There's a nice explanation for what unboxed 1-tuples do [on StackOverflow](https://stackoverflow.com/q/63441862). For this particular case the the idea is that allocations for `Black` and `White` are [shared](https://wiki.haskell.org/GHC/Memory_Footprint), so there never is a need to evaluate the results to WHNF.

Had a little back-and-forth about it [here](https://discourse.haskell.org/t/a-question-on-constructor-sharing-and-evaluation/9123), final STG seems to align with my expectations.